Design for warning on improper casing for PackageLicenseFile and PackageIcon

PathogenDavid opened the pull request:
The PR adds a design spec to address https://github.com/NuGet/Home/issues/9817

[Markdown view](https://github.com/PathogenDavid/NuGet-Home/blob/33f08de087d998d4041ef9c570737da68d024de2/designs/PackageValidation/FileNameCaseSensitivity.md)

PathogenDavid committed:
Added "Warn on improper license file name case" spec.

zivkan commented on designs/LicenseFileNameCaseSensitivity.md:
```diff
@@ -0,0 +1,63 @@
+
+# Warn on improper license file name case
```
This seems to be a pack validation, so I think it would be good to be in `designs/PackageValidation/`.

PathogenDavid replied:
Oops, I somehow didn't even notice the other folders. Moved in 2b7812b0e98437721a0187498b46790cd370926b

zivkan commented on designs/LicenseFileNameCaseSensitivity.md:
```diff
@@ -0,0 +1,63 @@
+
+# Warn on improper license file name case
```
I imagine the same problem exists for embedded icons, as the design is basically the same (nuspec property points to a file in the nupkg). I imagine the implementation could check for both at no higher cost than checking only one, so we should "expand" the scope.

PathogenDavid replied:
You're right, it does. I reworded the proposal to apply to package icons as well in https://github.com/NuGet/Home/pull/10049/commits/e19b3af5f919668d6712d3351f351a8d57f92e6b

![image](https://user-images.githubusercontent.com/278957/93663303-8bd94b00-fa2c-11ea-81c8-960e8ca85e69.png)

zivkan commented on designs/LicenseFileNameCaseSensitivity.md:
```diff
@@ -0,0 +1,63 @@
+
+## Solution
+
+Introduce a new warning when attempting to pack a package with improper casing on the `PackageLicenseFile`:
```
If we accept my proposal to make the warning apply to both embedded licenses as well as embedded icons, should we have one warning code per recourse type, or one warning code and make the message include sufficient information for the customer to understand where to fix the issue?

nkolev92 replied:
How about we keep the same error code, but slightly change the message?

And do it for both license + icon.

PathogenDavid replied:
If we don't re-use the same error codes like @nkolev92 is suggesting, I would imagine they should be separate error codes since [NU5030](https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu5030) and [NU5046](https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu5046) are already separate.

zivkan replied:
I forgot/didn't know about these existing errors. I very much like the idea of using the same codes and just changing the message when we detect the name with the wrong casing. 👍 

I don't want to add scope creep, but maybe we could also look for the filename in different directories in the nupkg, just in case the user didn't match the nupkg paths?

PathogenDavid replied:
The spec has been updated to use the existing error codes.

zivkan reviewed (approved):
I think this can be extended to work for both embedded icons, as well as licenses, at almost zero extra implementation cost. Otherwise I'm happy with the proposal.

nkolev92 reviewed (commented):
Thanks a lot for this @PathogenDavid. This is great. 

I think personally, I'd prefer this being an error with the same code as before, but an altered message. Let's pull in a few more people to get their thoughts!

Calling on @JonDouglas for help here! :)

PathogenDavid committed:
Updated file name case warning design to include package icons and consider error instead of warning.

PathogenDavid committed:
Moved LicenseFileNameCaseSensitivity into PackageValidation design folder and renamed it.

PathogenDavid committed:
Replaced dummy error code to clarify the warnings would not be the same if the existing error codes aren't reused.

PathogenDavid commented:
I've updated the design to apply to package icons after @zivkan's suggestion since they have the same issue. I also moved it to the `PackageValidation` folder.

I also added @nkolev92's suggestion to the open questions section to ensure it isn't forgotten. I only gravitated towards the new warning to avoid breaking people and since @zkat [showed preference for it](https://github.com/NuGet/Home/issues/9817#issuecomment-670076280).

Ignoring compat concerns I think it should definitely be an error. So if the NuGet team isn't worried about breaking old, malformed packages, I'm more than happy to go along the error route.

nkolev92 reviewed (approved):
This looks great so far, I think the only open question on NuGet side is warning or error. 

cc @JonDouglas @aortiz-msft 

Can we get someone to drive a quick internal discussion?

JonDouglas commented:
👍 to an error code. Given that if we warn for this scenario & a user can still be unsuccessful, an error is more proper.

zivkan commented on designs/PackageValidation/FileNameCaseSensitivity.md:
```diff
@@ -0,0 +1,74 @@
+
+## Open Questions
+
+Should this be an error that reuses the existing errors (NU5030 for licenses and NU5046 for icons) instead?
```
I saw yes, it's reuse the existing error codes, and add new messages if we can find a case-insensitive match.

PathogenDavid replied:
Spec updated.

zivkan reviewed (commented):
@PathogenDavid, given Jon's approval, let's close out the last open question, and get this spec merged! 👍

PathogenDavid committed:
Resolved open question about whether or not to make incorrect file casing an error.

PathogenDavid commented:
@zivkan Sorry for the wait, finally got around to updating the spec.

> I don't want to add scope creep, but maybe we could also look for the filename in different directories in the nupkg, just in case the user didn't match the nupkg paths?

(We never talked about this.) It wouldn't be the hardest thing to add, but I'm inclined to think this is probably an exceedingly rare situation and is out-of-scope. I added it to the considerations and labeled it as out of scope, but I'm willing to add it if you think it's common enough to warrant doing.

zivkan commented:
> Is there an easy way to run just a single test? Test Explorer didn't populate many of the tests for me

You're probably aware, but Test Explorer can take some time to finish discovering all test assemblies. I've never seen it only discover a subset of tests from within an assembly. But it "works for me".

From the command line you can go to the right project directory and run "dotnet test --filter YourTestNameHere". If you've already compiled the project in Visual Studio, or elsewhere, it's a bit faster to add `--no-build` or pass the path of the assembly to test itself. For example, in NuGet.Packaging.Test `dotnet test .\bin\Debug\netcoreapp5.0\NuGet.Packaging.Test.dll --filter Convention`.

And thank you for updating the spec! Merging now.

PathogenDavid commented:
> You're probably aware, but Test Explorer can take some time to finish discovering all test assemblies. I've never seen it only discover a subset of tests from within an assembly. But it "works for me".

Ah, you're right. I thought it had finished but it actually hadn't (and/or the test was nested down further than I expected.) For whatever reason it can't actually run tests though, but `dotnet test` works so I'll just use that.

Thanks for the merge! PR to NuGet.Client should be coming soon.

PathogenDavid commented:
I was going to wait until CI finished running, but it looks like CI doesn't automatically run on community PRs or something along those lines. The PR is here: https://github.com/NuGet/NuGet.Client/pull/3754

zivkan commented:
yes, that's right. See step 4 from our contributing guide: https://github.com/NuGet/NuGet.Client/blob/dev/CONTRIBUTING.md#contributing-step-by-step

We're making slow steps towards being able to run CI automatically, and make CI logs public, but we have quite a way to go yet.

PathogenDavid commented:
Ah, makes sense. I read that at one point but forgot that bit. Thanks for the confirmation!
